modules/300-prometheus/hooks/alertmanager_crd_discovery.go

```diff
@@ -112,122 +54,108 @@ func applyAlertmanagerCRDFilter(obj *unstructured.Unstructured) (go_hook.FilterR
-	deprecatedServices := handleDeperecatedAlertmanagerServices(input)
-	if len(deprecatedServices) > 0 {
-		// merge old - service discovery AlertManagers and new CR
-		for gr, values := range deprecatedServices {
-			if gr == "main" {
-				if _, ok := serviceDeclaredAlertmanagers["main"]; !ok {
-					serviceDeclaredAlertmanagers["main"] = make([]alertmanagerServiceInfo, 0)
-				}
-				serviceDeclaredAlertmanagers["main"] = append(serviceDeclaredAlertmanagers["main"], values...)
-			} else {
-				serviceDeclaredAlertmanagers[gr] = values
-			}
+			internalDeclaredAlertmanagers = append(internalDeclaredAlertmanagers, value)
 		}
 	}
 
-	input.Values.Set("prometheus.internal.alertmanagers", serviceDeclaredAlertmanagers)
+	input.Values.Set("prometheus.internal.alertmanagers.byAddress", addressDeclaredAlertmanagers)
+	input.Values.Set("prometheus.internal.alertmanagers.byService", serviceDeclaredAlertmanagers)
+	input.Values.Set("prometheus.internal.alertmanagers.internal", internalDeclaredAlertmanagers)
 
 	return nil
 }
 
-func handleDeperecatedAlertmanagerServices(input *go_hook.HookInput) map[string][]alertmanagerServiceInfo {
-	snaps := input.Snapshots["alertmanager_services"]
-	alertManagers := make(map[string][]alertmanagerServiceInfo)
-	for _, svc := range snaps {
-		alertManagerService := svc.(*alertmanagerService)
-
-		if _, ok := alertManagers[alertManagerService.Prometheus]; !ok {
-			alertManagers[alertManagerService.Prometheus] = make([]alertmanagerServiceInfo, 0)
-		}
-		alertManagers[alertManagerService.Prometheus] = append(alertManagers[alertManagerService.Prometheus], alertManagerService.Service)
-	}
-
-	return alertManagers
-}
-
-func parseServiceCR(am Alertmanager, snap []go_hook.FilterResult) (alertmanagerServiceInfo, error) {
-	var value alertmanagerServiceInfo
+func parseServiceCR(am Alertmanager, k8 k8s.Client) (alertmanagerService, error) {
+	var value alertmanagerService
 	serviceName, ok, err := unstructured.NestedString(am.Spec, "external", "service", "name")
 	if err != nil {
 		return value, err
 	}
 	if !ok {
-		return value, errors.New("service name required")
+		return value, fmt.Errorf("service name required: %v", am.Spec)
 	}
 
 	serviceNamespace, ok, err := unstructured.NestedString(am.Spec, "external", "service", "namespace")
 	if err != nil {
 		return value, err
 	}
 	if !ok {
-		return value, errors.New("service namespace required")
+		return value, fmt.Errorf("service namespace required: %v", am.Spec)
 	}
 
 	pathPrefix, ok, err := unstructured.NestedString(am.Spec, "external", "service", "path")
 	if err != nil || !ok {
 		pathPrefix = "/"
 	}
 
-	for _, s := range snap {
-		svc := s.(alertmanagerCRDService)
-		if svc.Name == serviceName && svc.Namespace == serviceNamespace {
-			value.Name = svc.Name
-			value.Namespace = svc.Namespace
-			value.Port = svc.Port
-			value.PathPrefix = pathPrefix
-			break
+	svc, err := k8.CoreV1().Services(serviceNamespace).Get(context.TODO(), serviceName, metav1.GetOptions{})
```

Comment:
Did not get the purpose of switching to the client-go there. Is it to avoid all service subscriptions?

Reply:
We discussed it internally and decided to use go client here to avoid the all services subscription.